### PR TITLE
[CLOUDSTACK-10240] ACS cannot migrate a local volume to shared storage

### PR DESCRIPTION
**Pull request by rafaelweingartner**
CloudStack is logically restricting the migration of local storages to shared storage and vice versa. This restriction is a logical one and can be removed for XenServer deployments. Therefore, we will enable migration of volumes between local-shared storages in XenServers independently of their service offering. This will work as an override mechanism to the disk offering used by volumes. If administrators want to migrate local volumes to a shared storage, they should be able to do so (the hypervisor already allows that). The same the other way around.

Extra information for reviewers:
This PS is introducing an overriding mechanism for volumes placement. We are providing a way for root administrators to override service/disk offering definitions by allocation volumes in storage pools that they could not have been allocated before if we followed their(volumes) service/disk offerings.
Therefore, it will not change the type of disk/service offerings when migrating volumes between local/shared pools. We will simply migrate them as requested by root admin. Furthermore, volumes will only be able to migrate to "suitable" storage pools. This means, storage pools that have same tags as the disk/service offering.

In summary, we are overriding placement considering location (shared/local) storage pools. However, we still consider storage tags to decide which storage pools are suitable or not.

### CODE REVIEW

File: api/src/main/java/org/apache/cloudstack/api/command/admin/storage/FindStoragePoolsForMigrationCmd.java [hunk 1]
```diff
@@ -29,14 +29,14 @@
...
 
 import com.cloud.storage.StoragePool;
 import com.cloud.utils.Pair;
 
-@APICommand(name = "findStoragePoolsForMigration", description = "Lists storage pools available for migration of a volume.", responseObject = StoragePoolResponse.class,
-        requestHasSensitiveInfo = false, responseHasSensitiveInfo = false)
+@APICommand(name = "findStoragePoolsForMigration", description = "Lists storage pools available for migration of a volume.", responseObject = StoragePoolResponse.class, requestHasSensitiveInfo = false, responseHasSensitiveInfo = false)
 public class FindStoragePoolsForMigrationCmd extends BaseListCmd {
-    public static final Logger s_logger = Logger.getLogger(FindStoragePoolsForMigrationCmd.class.getName());
+    public static final Logger s_logger = Logger.getLogger(FindStoragePoolsForMigrationCmdTest.class.getName());
```
[the remark this answers is not in the document]
Comment (rafaelweingartner) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/storage/FindStoragePoolsForMigrationCmd.java]: Thanks! Fixed.

[the remark this answers is not in the document]
Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/storage/FindStoragePoolsForMigrationCmd.java]: it is stil there, did you push?

[the remark this answers is not in the document]
Comment (rafaelweingartner) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/storage/FindStoragePoolsForMigrationCmd.java]: I forgot to push :(
my bad!

**Comment by rafaelweingartner** [comment]
@Daan, thanks for the feedback here!
I am starting to doubt that we are all using the same code formatter (I am using this one [1]). I can remove the code formatting changes and leave only the methods that I created if you prefer.

[1] https://cwiki.apache.org/confluence/download/attachments/29687985/ApacheCloudStack.xml?version=1&modificationDate=1382960469000&api=v2

**Comment by DaanHoogland** [comment]
@rafaelweingartner that is the only true formatter for cloudstack, isn't it. still i'd like to have reformattings in a different commit. the mindset of reviewing reformatting is a bit different than that of reviewing semantic/logic changes.

**Comment by rafaelweingartner** [comment]
I would say that there is only one formatting style for ACS. However, I have seen so many PRs like this one where we see a lot of formatting changes that I am starting to doubt if we are all using the same. 

I agree with you. My intention here was never to do code formatting. My eclipse is configured to format only edit lines, but for some reason it formatted all of the Java files I touched (maybe I pressed a ctrl+shift+f without noticing). Would you like me to remove the formatting and leave only the real changes I am introducing here?

**Comment by DaanHoogland** [comment]
no, never mind. (next time better!)

**Comment by rafaelweingartner** [comment]
ok @DaanHoogland.
Thanks!

**Comment by rafaelweingartner** [comment]
@DaanHoogland I separated the code changes from formatting one as your suggestion. I think it is way better now.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rafaelweingartner** [comment]
These failures are the same that appeared in other PR that is changing different things. How much can we trust these errors? Or an even better question, can we trust in a successful execution?

**Comment by DaanHoogland** [comment]
These are nested environments in which we test so they might be false positives as you imply by your question. I think we still need to check the errors. What PR shows these? I saw failures today but nowhere the exact same bunch, @rafaelweingartner ?

**Comment by rafaelweingartner** [comment]
The one I remember right now is the HA test, which I have seen failing in PRs 2438 and 2469. 
My doubt is the following, as long as we have false negatives (showing that there is a problem, when there isn't), can't we have false positive (showing that everything is ok, when it is not) as well?

**Comment by DaanHoogland** [comment]
@rafaelweingartner , first to clarify; my understanding of flase positive and false negitive is reversed from yours;
a false positive: a bug is reported but it isn't there
a false negative: a bug is not reported in spite of being there
kind of like in the doping control mafia; positive means your busted. maybe a bit counter intuitive but I'll run with it (pun as happy accident)

So more more to the point, a false positive might be due to a timing issue that occurs because of the nested nature of the test environment, and isn't a real issue. I can think of no false negative counterpart for that.
maybe we can think of more examples of the reverse but the issue is that we need to assess on each failure if it is noteworthy and we need to keep doing code reviews.

**Comment by rafaelweingartner** [comment]
@DaanHoogland thanks for the explanations.  Yes I was considering false positive/negative differently, but as long as we are able to explain them further as you did, we can understand each other.

So, it seems that most of these failures are due to runtime issues, normally from timeouts. Moreover, as far as we know it is not possible for the tests to return with a "success" message, when there is a problem. Unless, of course, if there is a problem/case that the current tests do not cover.

**Comment by DaanHoogland** [comment]
YES @rafaelweingartner , that is always the big question. I think we assume somethings there and to be sure we need to 'prove' each individual test. As soon as we've proven them we will start assuming again, unfortunately. How often do we need to revisit?

**Comment by rafaelweingartner** [comment]
That is a good question. For me, it is always best to start solving a bug/issue/new feature by unit tests because they cost less to run than these integration/smoke tests (this is what I have been doing so far). However, this requires a level of commitment (code has to be designed to be unit tested).

Regarding revisiting tests, I would check them every-time they fail. Otherwise, there is too much for too few hands.

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rafaelweingartner** [comment]
Now the failures increased... :)

**Comment by DaanHoogland** [comment]
:( jenkins had crashed due to disk full, should not be related and should not be. acting insane:
@blueorangutan test

**Comment by rafaelweingartner** [comment]
@DaanHoogland this test `test_hostha_enable_ha_when_host_in_maintenance` error seems to be persistent in other PRs as well. Can I consider that the tests finished successfully?

**Comment by rafaelweingartner** [comment]
@khos2ow would you mind reviewing this PR?

**Comment by DaanHoogland** [comment]
@rafaelweingartner , yes you can. It seems to be a timing issue. the host does not have a state needed in some occurences and does have that state in other runs.

**Comment by rafaelweingartner** [comment]
Thanks @DaanHoogland!

**Comment by rafaelweingartner** [comment]
If we have no objections I will be merging this one latter today.
